core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java

```diff
@@ -2145,4 +2150,37 @@ public void setTabletHostingGoal(String tableName, Range range, TabletHostingGoa
     }
   }
 
+  @Override
+  public Stream<HostingGoalForTablet> getTabletHostingGoal(final String tableName,
+      final Range range) throws TableNotFoundException {
+    EXISTING_TABLE_NAME.validate(tableName);
+
+    List<HostingGoalForTablet> hostingInfo = new ArrayList<>();
+
+    final Text scanRangeStart = (range.getStartKey() == null) ? null : range.getStartKey().getRow();
+    TableId tableId = context.getTableId(tableName);
+
+    try (TabletsMetadata m = context.getAmple().readTablets().forTable(tableId)
+        .overlapping(scanRangeStart, true, null).build()) {
+
+      for (TabletMetadata tm : m) {
+        final KeyExtent tabletExtent = tm.getExtent();
+        if (scanRangeStart != null && tm.getEndRow() != null
+            && tm.getEndRow().compareTo(scanRangeStart) < 0) {
+          // the end row of this tablet is before the start row, skip it
+          log.debug(">>>> tablet {} is before scan start range: {}", tabletExtent, scanRangeStart);
+          throw new RuntimeException("Bug in ample or this code.");
+        }
+        if (tm.getPrevEndRow() != null
+            && range.afterEndKey(new Key(tm.getPrevEndRow()).followingKey(PartialKey.ROW))) {
+          break;
+        }
+        hostingInfo
+            .add(new HostingGoalForTablet(new TabletIdImpl(tabletExtent), tm.getHostingGoal()));
+      }
+
+    }
```

Comment:
May be able to replace this loop with stream transformations and avoid buffering all of the tablets in memory. Can use the peek() method on stream to check for the bug the loop is checking for. Can use the takeWhile() method on stream to take tablets until they are out range.

If using a stream, then can not use a try with resources in this method. Can not close the scanner and return a stream that is reading from the scanner. Could instead call the streams onClose() method and pass a lambda that closes the TabletsMetadata object. Would need to to add javadoc encouraging user of this method to close the returned stream.

Reply:
@keith-turner can you take a look at this commit and see if it addresses your comments. My stream knowledge is not deep so let me know if additional changes or another direction needs to be taken.

Reply:
@jmark99 it looks good.